kernel/tests/golden_tables.rs

```diff
@@ -0,0 +1,343 @@
+            if let Some(mut mask) = sr.mask {
+                let extra_rows = record_batch.num_rows() - mask.len();
+                if extra_rows > 0 {
+                    // we need to extend the mask here in case it's too short
+                    mask.extend(std::iter::repeat(true).take(extra_rows));
+                }
+                filter_record_batch(&record_batch, &mask.into()).unwrap()
+            } else {
+                record_batch
+            }
+        })
+        .collect();
+
+    let expected = read_expected(&expected_path).await.unwrap().unwrap();
+    let schema: ArrowSchemaRef = Arc::new(scan.schema().as_ref().try_into().unwrap());
+
+    // convert the batch +00:00 to UTC
+    let result: Vec<RecordBatch> = batches
+        .iter()
+        .map(|batch| {
+            let result: Vec<ArrayRef> = batch.columns().iter().map(normalize_col).collect();
+            RecordBatch::try_new(schema.clone(), result).unwrap()
+        })
+        .collect();
+
+    let result = concat_batches(&schema, &result).unwrap();
+    assert_batches_eq!(expected, result);
+    Ok(())
+}
+
+macro_rules! full_scan_test {
+    ($test_name:literal) => {
+        paste! {
+            #[tokio::test]
+            async fn [<golden_ $test_name:snake>]() {
+                latest_snapshot_test($test_name).await.unwrap();
+            }
+        }
+    };
+}
+
+// same as golden_test but we expect the test to fail
+// TODO: check actual error
+macro_rules! negative_test {
+    ($test_name:literal) => {
+        paste! {
+            #[tokio::test]
+            #[should_panic]
+            async fn [<golden_negative_ $test_name:snake>]() {
+                latest_snapshot_test($test_name).await.unwrap();
+            }
+        }
+    };
+}
+
+macro_rules! skip_test {
+    ($test_name:literal: $reason:literal) => {
+        paste! {
+            #[ignore = $reason]
+            #[tokio::test]
+            async fn [<golden_skip_ $test_name:snake>]() {
+                latest_snapshot_test($test_name).await.unwrap();
+            }
+        }
+    };
+}
+
+macro_rules! golden_test {
+    ($test_name:literal, $test_fn:expr) => {
+        paste! {
+            #[tokio::test]
+            async fn [<golden_ $test_name:snake>]() -> Result<(), Box<dyn std::error::Error>> {
+                let test_dir = load_test_data($test_name).unwrap();
+                let test_path = test_dir.path().join($test_name);
+                let table_path = std::fs::canonicalize(test_path.join("table")).unwrap();
+                let table_path = url::Url::from_directory_path(table_path).unwrap();
+                let engine =
+                    DefaultEngine::try_new(
+                        &table_path,
+                        std::iter::empty::<(&str, &str)>(),
+                        Arc::new(TokioBackgroundExecutor::new()),
+                    ).unwrap();
+                let table = Table::new(table_path);
```

Comment:
```suggestion
                let table = Table::try_from_uri(&table_path)?;
                let engine =
                    DefaultEngine::try_new(
                        table.location(),
                        std::iter::empty::<(&str, &str)>(),
                        Arc::new(TokioBackgroundExecutor::new()),
                    ).unwrap();
```

You likely don't need to canonicalize it either. making the table should do that.

Reply:
needed to just do to_str and worked!